plugins/dig-now.cpp

```diff
@@ -0,0 +1,870 @@
+static void clean_ramps(MapExtras::MapCache &map, const DFCoord &pos) {
+    clean_ramp(map, pos);
+    clean_ramp(map, DFCoord(pos.x-1, pos.y, pos.z));
+    clean_ramp(map, DFCoord(pos.x+1, pos.y, pos.z));
+    clean_ramp(map, DFCoord(pos.x, pos.y-1, pos.z));
+    clean_ramp(map, DFCoord(pos.x, pos.y+1, pos.z));
+}
+
+// destroys any colonies located at pos
+static void destroy_colony(const DFCoord &pos) {
+    auto same_pos = [&](df::vermin *colony){ return colony->pos == pos; };
+
+    auto &colonies = world->vermin.colonies;
+    auto found_colony = std::find_if(begin(colonies), end(colonies), same_pos);
+    if (found_colony == end(colonies))
+        return;
+    colonies.erase(found_colony);
+
+    auto &all_vermin = world->vermin.all;
+    all_vermin.erase(
+        std::find_if(begin(all_vermin), end(all_vermin), same_pos));
+}
+
+struct dug_tile_info {
+    DFCoord pos;
+    df::tiletype_material tmat;
+    df::item_type itype;
+    int32_t imat; // mat idx of boulder/gem potentially generated at this pos
+
+    dug_tile_info(MapExtras::MapCache &map, const DFCoord &pos) {
+        this->pos = pos;
+
+        df::tiletype tt = map.tiletypeAt(pos);
+        tmat = tileMaterial(tt);
+
+        switch (map.BlockAtTile(pos)->veinTypeAt(pos)) {
+            case df::inclusion_type::CLUSTER_ONE:
+            case df::inclusion_type::CLUSTER_SMALL:
+                itype = df::item_type::ROUGH;
+                break;
+            default:
+                itype = df::item_type::BOULDER;
+        }
+
+        imat = -1;
+        if (tileShape(tt) == df::tiletype_shape::WALL
+                && (tmat == df::tiletype_material::STONE
+                    || tmat == df::tiletype_material::MINERAL
+                    || tmat == df::tiletype_material::FEATURE))
+            imat = map.baseMaterialAt(pos).mat_index;
+    }
+};
+
+static bool is_diggable(MapExtras::MapCache &map, const DFCoord &pos,
+                        df::tiletype tt) {
+    df::tiletype_material mat = tileMaterial(tt);
+    switch (mat) {
+    case df::tiletype_material::CONSTRUCTION:
+    case df::tiletype_material::POOL:
+    case df::tiletype_material::RIVER:
+    case df::tiletype_material::TREE:
+    case df::tiletype_material::ROOT:
+    case df::tiletype_material::LAVA_STONE:
+    case df::tiletype_material::MAGMA:
+    case df::tiletype_material::HFS:
+    case df::tiletype_material::UNDERWORLD_GATE:
+        return false;
+    default:
+        break;
+    }
+
+    if (mat == df::tiletype_material::FEATURE) {
+        // adamantine is the only is diggable feature
+        t_feature feature;
+        return map.BlockAtTile(pos)->GetLocalFeature(&feature)
+                && feature.type == feature_type::deep_special_tube;
+    }
+
+    return true;
+}
+
+static bool dig_tile(color_ostream &out, MapExtras::MapCache &map,
+                     const DFCoord &pos, df::tile_dig_designation designation,
+                     std::vector<dug_tile_info> &dug_tiles) {
```

Comment:
@myk002 How would you feel about an option where instead of a cached map it's the real thing?

Reply:
As much as I dislike having to use MapCache, it has some incredibly useful functionality around getting and setting tile material. And once I use it for anything, I need to use it for everything since the MapCache state isn't synced back to the real map until you call WriteAll, and if we do that on every change the performance would be dismal.